### PR TITLE
SpotLightMap v3

### PR DESCRIPTION
**Pull request by mbredif**
rebase of #20290

### CODE REVIEW

**Review by Mugen87 — APPROVED** [review]
The PR seems to work fine! I've just simplified `webgl_lights_spotlight` a bit. 

However, I do not vote to change `webgl_shadowmap_viewer`. IMO, the additional light with its camera helper make the visuals crowded. @mrdoob Please double check^^.

**Comment by mbredif** [comment]
I agree that examples should be simplified (removing the video from webgl_lights_spotlight and leaving webgl_shadowmap_viewer untouched). They were there only to show/test the PR.

**Comment by mrdoob** [comment]
> However, I do not vote to change `webgl_shadowmap_viewer`. IMO, the additional light with its camera helper make the visuals crowded. @mrdoob Please double check^^.

Agreed 👍

**Comment by Mugen87** [comment]
Okay, reverted the changes to `webgl_shadowmap_viewer`.

**Comment by Deenis** [comment]
This is a long waited feature for me! Great job! However, is it possible to keep aspect ratio of texture, or similar? I'm projecting a 16/9 video, but the projected texture is stretched. Neither wrapping, scaling or adjusting height / width of the texture, seems to have any effect.

Pardon me, if this is a stupid question. I'm not that experienced with ThreeJS.

**Comment by crubier** [comment]
Let's get this done 🙏 😭 Are there any blockers at the moment ?

**Comment by inear** [comment]
Is this PR under consideration to be included in the future, or is the milestones automatically increased? Would love this feature :) Wish I could help, but would just mess things up.

File: src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js [hunk 1]
```diff
@@ -68,9 +72,29 @@ IncidentLight directLight;
...
+			inSpotLightMap = inSpotLightMap && ( spotColor.a > 0. );
+			directLight.visible = directLight.visible || inSpotLightMap;
+			directLight.color = inSpotLightMap ? mix( directLight.color, spotLight.color * spotColor.rgb, spotColor.a ) : directLight.color;
+		#endif
+
+		#undef SPOT_LIGHT_MAP_INDEX
+
+		#if defined(USE_SHADOWMAP) && ( UNROLLED_LOOP_INDEX < NUM_SPOT_LIGHT_SHADOWS )
+			spotLightShadow = spotLightShadows[ i ];
+			directLight.color *= all( bvec2( directLight.visible, receiveShadow ) ) ? getShadow( spotShadowMap[ i ], spotLightShadow.shadowMapSize, spotLightShadow.shadowBias, spotLightShadow.shadowRadius, vSpotLightCoord[ i ] ) : 1.0;
```
Comment (LeviPesin) [inline src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js:97]: `directLight.visible && receiveShadow` is better than `all( bvec2( directLight.visible, receiveShadow ) )`.

**Comment by AdaRoseCannon** [comment]
this would be really useful for me for an upcoming project, I would be happy to handle the merge conflicts

**Comment by usefulthink** [comment]
Unless @mbredif is up to that, I would like to rebase this so it can finally land for a future versions. 
I checked and the overall conflicts are really minimal, but doing a proper rebase of the individual comments would be pretty painful. 

Would it be ok for me to squash rebase this and push the new commit with proper `Co-authored-by` attributions? Or does anyone know a good strategy to keep the individual commits intact for a rebase?

What do you think @Mugen87?

**Comment by Mugen87** [comment]
> Would it be ok for me to squash rebase this and push the new commit with proper Co-authored-by attributions?

That sounds good to me.

**Comment by mbredif** [comment]
@usefulthink @Mugen87 , I just rebased this PR !

File: src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js (hunk 1 above)
Comment (mbredif) [reply src/renderers/shaders/ShaderChunk/lights_fragment_begin.glsl.js:122]: Maybe, I am just following the same pattern as the other tests in this file.

**Comment by mrdoob** [comment]
Thanks! Sorry that it took... 4 years... 🫣

**Comment by mrdoob** [comment]
After playing a bit with it, I find it strange that `penumbra` doesn't interact with it...

Screen Shot 2022-08-24 at 9 28 41 AM

Also, if you set `penumbra` to `0` you can see the spotlight and the map on top:

Screen Shot 2022-08-24 at 9 30 38 AM

Makes me wonder if `SpotLight` is the right place for this... Maybe we should create `RectLight`/`RectangularLight`?

**Comment by mbredif** [comment]
That is a matter of taste

* `mix( directLight.color, spotLight.color * spotColor.rgb, spotColor.a )` is how it is implemented : the spot texture alpha mixes the cone of penumbra and the rgb values of the texture.

* `directLight.color * spotLight.color * spotColor.rgb` is maybe the kind of modulation you have in mind (alpha is unused), however that modulation does not allow to deactivate entirely the penumbra (which can now be done with a texture with alpha=1)

image

**Comment by mrdoob** [comment]
@WestLangley What do you think?

**Comment by WestLangley** [comment]
>directLight.color * spotLight.color * spotColor.rgb is maybe the kind of modulation you have in mind

Do you mean
```js
directLight.color = spotLight.color * spotColor.rgb
```

**Comment by mbredif** [comment]
I gave a try in #24539 to enable the use of spotlight textures with and without the cone of light effect. Let's switch the discussion there :)

**Pull request merged**